waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,605 @@
+        edn-opts {:readers {'waiter/param params
+                            'waiter/param-str (comp str params)
+                            'waiter/port #(+ port0 %)
+                            'waiter.fn/into (fn [[xs ys]] (into xs ys))
+                            'waiter.fn/lower-case string/upper-case
+                            'waiter.fn/str #(apply str %)
+                            'waiter.fn/upper-case string/upper-case}}]
+    (try
+      (->> rs-spec-file-path slurp (edn/read-string edn-opts))
+      (catch Throwable e
+        (log/error e "Error creating ReplicaSet specification for" service-id)
+        (throw e)))))
+
+(defn- create-service
+  "Reify a Waiter Service as a Kuberentes ReplicaSet."
+  [service-id descriptor {:keys [api-server-url http-client] :as scheduler} service-id->password-fn]
+  (let [{:strs [run-as-user] :as service-description} (:service-description descriptor)
+        spec-json (service-spec scheduler service-id service-description service-id->password-fn)
+        request-url (str api-server-url
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (service-description->namespace service-description)
+                         "/replicasets")
+        response-json (api-request http-client request-url
+                                   :body (as-json spec-json)
+                                   :request-method :post)]
+    (when-not (= "ReplicaSet" (:kind response-json))
+      (log/error "Invalid response from ReplicaSet create request"
+                 (as-json response-json)))
+    (replicaset->Service response-json)))
+
+(defn- delete-service
+  "Delete the Kubernetes ReplicaSet corresponding to a Waiter Service.
+   Ensures that all controlled Pods are deleted before removing the ReplicaSet.
+   This operation will leave Waiter in a sane state on a failure."
+  [{:keys [api-server-url http-client] :as scheduler} service]
+  (when-not service
+    (log/error "Null service passed to kubernetes delete-service.")
+    (ss/throw+ {:status 404 :message "Service not found"}))
+  (let [replicaset-url (str api-server-url
+                            "/apis/extensions/v1beta1/namespaces/"
+                            (:namespace service)
+                            "/replicasets/"
+                            (:k8s-name service))]
+    (patch-object-json replicaset-url http-client
+                       [{:op :replace :path "/metadata/annotations/waiter~1app-status" :value "killed"}
+                        {:op :replace :path "/spec/replicas" :value 0}])
+    (doseq [pod (get-replicaset-pods scheduler service)
+            :let [pod-url (->> pod :metadata :selfLink (str api-server-url))]]
+      (api-request http-client pod-url :request-method :delete))
+    (api-request http-client replicaset-url :request-method :delete)
+    {:message (str "Kubernetes deleted ReplicaSet for " (:k8s-name service))
+     :result :deleted}))
+
+(defn service-id->service
```

Comment:
Looks like most places this is invoked we are mainly interested in the metadata associated with the service (`id`, `namespace`, `pod-name` etc). For those call sites, we can then have a separate function and rely on a cached version of this information of a new data structure and pass it around? This will avoid having to make the k8s api call over the wire to build the metadata.

Reply:
I agree, but I'll leave that optimization for a future PR. We'll be adding local caching of state when I add a watcher to get pushed state updates from the Kubernetes master, and I think this would follow naturally from that refactoring.